base62.go

```diff
@@ -25,28 +25,54 @@ func base2base(in []byte, inBase int, outBase int) []byte {
 		}
 
 		// this is really a prepend or remainder to result
-		res = append([]byte{byte(remainder)}, res...)
+		dst = append(dst, byte(remainder))
 		bs = quotient
 	}
 
-	return res
+	reverse(dst)
+	return dst
 }
 
-func encodeBase62(in []byte) string {
-	encoded := base2base(in, 256, 62)
-	for i, b := range encoded {
-		encoded[i] = base62Characters[b]
+func base2base(src []byte, inBase int, outBase int) []byte {
+	return appendBase2Base(nil, src, inBase, outBase)
+}
+
+func appendEncodeBase62(dst []byte, src []byte) []byte {
```

Comment:
Changing this to work on []byte + adding the []byte Append function to the external interface is actually a pretty big change. This is because this patch makes a subtle semantic modification: the base2base function is purely mathematical, so it works on integer data. In the existing code, the transformation from the integer form to characters in string form abides by the semantic rules of strings. bytestrings don't have encoding metadata, where-as a string does (always UTF-8 in Go!). A rune is not semantically the same as a byte, and a string is not semantically the same as a []byte. 

Even though it's convenient that base62 happens to fit into the []byte range of the UTF-8 character set, this causes an _implicit_ coupling between the external Append() method, it's users, and the internal library functions. The problem is pretty benign internally, but much more acute externally as it "leaks" this implicit coupling out through Append(). The only way to use the []byte returned from Append() is as a UTF-8 encoded string, but this is entirely implicit and there are no guard rails.

Thus this long-winded explanation means that I am -1 on exposing a UTF-8 string as []byte through the external interface. Use cases which are performance-sensitive should use the binary representation. The binary representation will _always_ be MUCH faster than the encoded version. I could be swayed from this if we can point to an existing convention/pattern within Go where passing strings as []byte outside of cases that are (semantically) arbitrary bytestrings in the first place.